### PR TITLE
Test to configure rbd compression hint settings at multiple levels

### PR DESCRIPTION
**Pull request by sunilangadi2**
# Description
Automated below test case: Test to verify "rbd_compression_hint" configs in globally, per-pool, or per-image
[CEPH-83574644](https://polarion.engineering.redhat.com/polarion/#/project/CEPH/workitem?id=CEPH-83574644)

Test flow:
1. Create a pool and an Image, write some data on it
2. set bluestore_compression_mode to passive to enable rbd_compression_hint feature
3. verify "rbd_compression_hint" to "compressible" on global, pool and image level
4. verify "rbd_compression_hint" to "incompressible" on global, pool and image level
5. Repeat the above steps for ecpool

Test result:
[Passed] http://magna002.ceph.redhat.com/cephci-jenkins/cephci-run-46AKM3

Please include Automation development guidelines. Source of Test case - New Feature/Regression Test/Close loop of customer BZs
<details>

<summary>click to expand checklist</summary>

- [ ] Create a test case in Polarion reviewed and approved.
- [ ] Create a design/automation approach doc. Optional for tests with similar tests already automated.
- [ ] Review the automation design
- [ ] Implement the test script and perform test runs
- [ ] Submit PR for code review and approve
- [ ] Update Polarion Test with Automation script details and update automation fields
- [ ] If automation is part of Close loop, update BZ flag qe-test_coverage “+” and link Polarion test
</details>

### CODE REVIEW

File: tests/rbd/test_rbd_compression.py [hunk 1]
```diff
@@ -0,0 +1,133 @@
...
+    """
+    Test for rbd_compression_hint settings.
+    Args:
+        **kw: test data
+
+    Returns:
+        int: The return value. 0 for success, 1 for failure.
+    """
+    log.info(
+        "Starting CEPH-83574644 - Validate rbd_compression_hint config"
```
Comment (rkhambad) [inline tests/rbd/test_rbd_compression.py:116]: give a space after config.

File: tests/rbd/test_rbd_compression.py [hunk 2]
```diff
@@ -0,0 +1,133 @@
...
+                "osd", "bluestore_compression_mode", output=True
+            ).strip()
+            != "passive"
+        ):
+            log.error("Not able to set bluestore_compression_mode to passive")
+            return 1
+
+        log.info("Able to set bluestore_compression_mode to passive")
+
+        for option in ["compressible", "incompressible"]:
```
Comment (rkhambad) [inline tests/rbd/test_rbd_compression.py:60]: [optional] we are verifying set and get twice on global level, we can think to run on global only once.

File: tests/rbd/rbd_utils.py [hunk 3]
```diff
@@ -773,6 +773,35 @@ def set_config(self, level, entity, key, value, **kw):
         """
         return self.exec_cmd(cmd=f"rbd config {level} set {entity} {key} {value}", **kw)
 
+    def set_ceph_config(self, entity, key, value, **kw):
```
Comment (rkhambad) [inline tests/rbd/rbd_utils.py:776]: [optional] we can think of modifying the existing set_config and get_config method, according to the parameter passed in function.

Comment (sunilangadi2) [reply tests/rbd/rbd_utils.py:776]: i feel it's good to keep `ceph` commands separate over `rbd` commands hence added this method.

File: tests/rbd/test_rbd_compression.py (hunk 2 above)
Comment (sunilangadi2) [reply tests/rbd/test_rbd_compression.py:134]: I don't think so, looks like you got confused with `entity` global level has entity of type global.

File: tests/rbd/test_rbd_compression.py (hunk 1 above)
Comment (sunilangadi2) [reply tests/rbd/test_rbd_compression.py]: Done

**Comment by sunilangadi2** [comment]
Added validation steps at each config level,
Test result:[Passed]
http://magna002.ceph.redhat.com/cephci-jenkins/cephci-run-6MUQGX/

File: tests/rbd/rbd_utils.py (hunk 3 above)
Comment (manasagowri) [reply tests/rbd/rbd_utils.py:776]: I agree with Sunil here, let's not combine rbd config and ceph config.

Comment (manasagowri) [reply tests/rbd/rbd_utils.py:776]: But @sunilangadi2 set_ceph_config and get_ceph_config are generic methods not related to RBD object, so they can be moved out of the RBD class.

File: tests/rbd/rbd_utils.py [hunk 4]
```diff
@@ -773,6 +773,35 @@ def set_config(self, level, entity, key, value, **kw):
...
+        ceph config get <entity> <key>
+
+        Args:
+            entity: config entity (mon, osd)
+            key: config option key (ex., rbd_move_to_trash_on_remove)
+            kw: other ceph arguments like config options
+        Returns:
+            exec_cmd response
+        """
+        return self.exec_cmd(cmd=f"ceph config get {entity} {key}", **kw)
```
Comment (manasagowri) [inline tests/rbd/rbd_utils.py:803]: Same, please move this method out of the RBD class.

File: tests/rbd/test_rbd_compression.py [hunk 5]
```diff
@@ -0,0 +1,222 @@
...
+    Args:
+        mode: type of compression hints (compressible or incompressible)
+        image_spec: poolname + imagename
+        **kw: test data
+
+    Returns:
+        int: The return value. 0 for success, 1 for failure.
+    """
+    config = kw.get("config")
+    io = config.get("io_total")
```
Comment (manasagowri) [inline tests/rbd/test_rbd_compression.py:40]: Please add default values, if config is not passed by user.

File: tests/rbd/test_rbd_compression.py [hunk 6]
```diff
@@ -0,0 +1,222 @@
...
+    out = self.exec_cmd(cmd="ceph df detail --format json-pretty", output=True)
+    output = json.loads(out.rstrip())
+    pool_info = output["pools"]
+
+    # Iterate over the pools to get compressed data information
+    for pool in pool_info:
+        if pool["name"] == pool_name:
+            compressed_data = pool["stats"]["compress_bytes_used"]
+            break
+    log.debug(pool["stats"])
```
Comment (manasagowri) [inline tests/rbd/test_rbd_compression.py:57]: I think this log.debug should be inside the for loop isn't it?

File: tests/rbd/test_rbd_compression.py [hunk 7]
```diff
@@ -0,0 +1,222 @@
...
+    if mode == "compressible":
+        if compressed_data == 0:
```
Comment (manasagowri) [inline tests/rbd/test_rbd_compression.py:59]: Are you sure its always 0 and not empty or something?

File: tests/rbd/test_rbd_compression.py [hunk 8]
```diff
@@ -0,0 +1,222 @@
...
+        if (
+            rbd.get_ceph_config(
+                "osd", "bluestore_compression_mode", output=True
+            ).strip()
+            != "passive"
+        ):
+            log.error("Not able to set bluestore_compression_mode to passive")
+            return 1
+
+        log.info("Able to set bluestore_compression_mode to passive")
```
Comment (manasagowri) [inline tests/rbd/test_rbd_compression.py:122]: The sentence could be "bluestore_compression_mode set to passive"

**Review by manasagowri — CHANGES_REQUESTED** [review]

File: tests/rbd/test_rbd_compression.py (hunk 6 above)
Comment (VasishtaShastry) [reply tests/rbd/test_rbd_compression.py:57]: @sunilangadi2 Also, please add more info to debug message, just printing numbers wouldn't help

File: tests/rbd/test_rbd_compression.py [hunk 9]
```diff
@@ -0,0 +1,222 @@
...
+    compression_mode = config.get("compression_mode")
+    compression_ratio = config.get("compression_ratio")
+
+    pool_name = config[pool_type]["pool"]
+
+    compression_configs = {
+        "compression_algorithm": compression_algorithm,
+        "compression_mode": compression_mode,
+        "compression_required_ratio": compression_ratio,
+    }
```
Comment (VasishtaShastry) [inline tests/rbd/test_rbd_compression.py:107]: If we are not using compression_ratio/compression_mode/compression_algorithm after using it to initiate compression_configs, why do we need them ?
In other words, can we merge 
line 97-99 with 104-106

**Review by VasishtaShastry — CHANGES_REQUESTED** [review]

File: tests/rbd/test_rbd_compression.py (hunk 7 above)
Comment (sunilangadi2) [reply tests/rbd/test_rbd_compression.py:62]: yes it's always 0 only for uncompressed data refer `'compress_bytes_used': 0`

File: tests/rbd/rbd_utils.py (hunk 3 above)
Comment (sunilangadi2) [reply tests/rbd/rbd_utils.py]: Done

File: tests/rbd/rbd_utils.py (hunk 4 above)
Comment (sunilangadi2) [reply tests/rbd/rbd_utils.py]: Done

File: tests/rbd/test_rbd_compression.py (hunk 5 above)
Comment (sunilangadi2) [reply tests/rbd/test_rbd_compression.py]: done

File: tests/rbd/test_rbd_compression.py (hunk 6 above)
Comment (sunilangadi2) [reply tests/rbd/test_rbd_compression.py]: done added

File: tests/rbd/test_rbd_compression.py (hunk 8 above)
Comment (sunilangadi2) [reply tests/rbd/test_rbd_compression.py]: modified accordingly

File: tests/rbd/rbd_utils.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (sunilangadi2) [reply tests/rbd/rbd_utils.py]: done

File: tests/rbd/test_rbd_compression.py [hunk 10]
```diff
@@ -0,0 +1,133 @@
+"""Test case covered -
+    CEPH-83574644 - Validate "rbd_compression_hint" config
+    settings on globally, Pool level, and image level.
+
+    Pre-requisites :
+    1. Cluster must be up and running with capacity to create pool
+       (At least with 64 pgs)
```
[the remark this answers is not in the document]
Comment (sunilangadi2) [reply tests/rbd/test_rbd_compression.py]: removed pg info

File: tests/rbd/test_rbd_compression.py (hunk 9 above)
Comment (sunilangadi2) [reply tests/rbd/test_rbd_compression.py]: thanks for the code optimization point, Done

**Comment by sunilangadi2** [comment]
thanks for the review all addressed all review comments along with merge conflicts,
Test result with latest updated code
[Passed] http://magna002.ceph.redhat.com/cephci-jenkins/cephci-run-IJ7V4G/
Please review it.

File: tests/rbd/test_rbd_compression.py (hunk 6 above)
Comment (VasishtaShastry) [reply tests/rbd/test_rbd_compression.py]: okay with log message

**Review by VasishtaShastry — APPROVED** [review]

**Comment by sunilangadi2** [comment]
Due to some missing of sequnec of commands during commit 
this PR related commits got messed up 
Raised the new PR with test automation code changes related to it here
https://github.com/red-hat-storage/cephci/pull/2575

Hence closing this PR.

**Pull request closed without merging**